delta-proto/src/main/java/io/cdap/delta/proto/Stage.java

```diff
@@ -38,6 +49,113 @@ public Plugin getPlugin() {
+   *    If new plugin is in range, do not change range. (Note: It would not change range even though new plugin is in
+   *    different scope).
+   *
+   * @param updateContext To use helper functions like getPluginArtifacts.
+   * @return Updated plugin object to be used for the udated stage. Returned null if no changes to current plugin.
+   */
+  private Plugin upgradePlugin(ApplicationUpdateContext updateContext) throws Exception {
+    // Find the plugin with max version from available candidates.
+    Optional<ArtifactId> newPluginCandidate =
+      updateContext.getPluginArtifacts(plugin.getType(), plugin.getName(), null).stream()
+        .max(Comparator.comparing(ArtifactId::getVersion));
+    if (!newPluginCandidate.isPresent()) {
+      // This should not happen as there should be at least one plugin candidate same as current.
+      // TODO: Consider throwing exception here.
+      return plugin;
+    }
+
+    ArtifactId newPlugin = newPluginCandidate.get();
+    String newVersion = getUpgradedVersionString(newPlugin);
+    // If getUpgradedVersionString returns null, candidate plugin is not valid for upgrade.
+    if (newVersion == null) {
+      return plugin;
+    }
+
+    Artifact newArtifact = new Artifact(newPlugin.getScope().name(), newPlugin.getName(), newVersion);
+
+    return new Plugin(plugin.getName(), plugin.getType(), plugin.getProperties(), newArtifact);
+  }
+
+  /**
+   * Returns new valid version string for plugin upgrade if any changes are required. Returns null if no change to
+   * current plugin version.
+   * Artifact selector config only stores plugin version as string, it can be either fixed version or range.
+   * Hence, if the plugin version is fixed, replace the fixed version with newer fixed version. If it is a range,
+   * move the upper bound of the range to the newest version.
+   *
+   * @param newPlugin New candidate plugin for updating plugin artifact.
+   * @return version string to be used for new plugin. Might be fixed version/version range string depending on
+   *         current use.
+   */
+  @Nullable
+  private String getUpgradedVersionString(ArtifactId newPlugin) {
+    ArtifactVersionRange currentVersionRange;
+    try {
+      currentVersionRange =
+        io.cdap.cdap.api.artifact.ArtifactVersionRange.parse(Objects.requireNonNull(plugin.getArtifact().getVersion()));
+    } catch (Exception e) {
+      LOG.warn("Issue in parsing version string for plugin {}, ignoring stage {} for upgrade.", plugin, name, e);
+      return null;
+    }
+
+    if (currentVersionRange.isExactVersion()) {
+      if (currentVersionRange.getLower().compareTo(newPlugin.getVersion()) < 0) {
+        // Current version is a fixed version and new version is higher than current.
+        return newPlugin.getVersion().getVersion();
+      }
+      return null;
+    }
+
+    if (!currentVersionRange.isExactVersion()) {
```

Comment:
seems like this will always be true at this point

Reply:
good catch! fixed.